crates/sui-source-validation/src/lib.rs

```diff
@@ -396,3 +474,273 @@ fn local_modules(
+            flavor: Flavor::Sui,      /* does not matter, unused for current_toolchain */
+        };
+
+        if sui_types::is_system_package(local_unit.unit.address.into_inner()) {
+            // System packages are always compiled with the current compiler.
+            package_version_map.insert(*package, current_toolchain);
+            continue;
+        }
+
+        let package_root = SourcePackageLayout::try_find_root(&local_unit.source_path)?;
+        let lock_file = package_root.join("Move.lock");
+        if !lock_file.exists() {
+            // No lock file implies current compiler for this package.
+            package_version_map.insert(*package, current_toolchain);
+            continue;
+        }
+
+        let mut lock_file = File::open(lock_file)?;
+        let toolchain_version = ToolchainVersion::read(&mut lock_file)?;
+        lock_file.rewind()?;
+        let lock_version = Header::read(&mut lock_file)?.version;
+        match toolchain_version {
+            // No ToolchainVersion and old Move.lock version implies legacy compiler.
+            None if lock_version == LEGACY_MOVE_LOCK_VERSION => {
+                debug!("{package} on legacy compiler",);
+                let legacy_toolchain = ToolchainVersion {
+                    compiler_version: LEGACY_COMPILER_VERSION.into(),
+                    edition: Edition::LEGACY,
+                    flavor: Flavor::Sui,
+                };
+                package_version_map.insert(*package, legacy_toolchain);
+            }
+            // No ToolchainVersion and new Move.lock version implies current compiler.
+            None => {
+                debug!("{package} on current compiler @ {CURRENT_COMPILER_VERSION}",);
+                package_version_map.insert(*package, current_toolchain);
+            }
+            // This dependency uses the current compiler.
+            Some(ToolchainVersion {
+                compiler_version, ..
+            }) if compiler_version == CURRENT_COMPILER_VERSION => {
+                debug!("{package} on current compiler @ {CURRENT_COMPILER_VERSION}",);
+                package_version_map.insert(*package, current_toolchain);
+            }
+            // This dependency needs a prior compiler. Mark it and compile.
+            Some(toolchain_version) => {
+                println!(
+                    "{} {package} compiler @ {}",
+                    "REQUIRE".bold().green(),
+                    toolchain_version.compiler_version.yellow(),
+                );
+                package_version_map.insert(*package, toolchain_version);
+            }
+        }
+    }
+
+    let mut units = vec![];
+    let mut package_compiled: HashSet<Symbol> = HashSet::new();
+    // Iterate over compiled units, and determine whether they need to be recompiled and replaced by a prior compiler.
+    for (package, local_unit) in compiled_units {
+        match package_version_map.get(package) {
```

Comment:
It seems like the interchange between the earlier loop and this loop could be made cleaner by stashing the `local_unit` in the map? That way we can leverage the checks in the earlier loop to guarantee that we only process each `package` once, and are guaranteed to have a toolchain version to read for it?

Reply:
05445f1260604c3550132b57137e76a825a216ec does what I _think_ we're shooting for and gives those guarantees